Bjorncs/sia backed http client

bjorncs opened the pull request:
Introduce a wrapper for Apache httpclient to be used in clients that requires credentials from a ServiceIdentityProvider. Does not rely on the listener interface in ServiceIdentityProvider.

FYI @freva

bjorncs committed:
Add http client backed by Apache httpclient + ServiceIdentityProvider

bjorncs committed:
Use SiaBackedApacheHttpClient in DefaultZtsClient

bjorncs committed:
Fix race conditions by using the global lock

Replace AtomicInteger with int and use clientLock to synchronize all
access to refer() and release(). Remove synchronized from constructor.

hakonhall commented on vespa-athenz/src/main/java/com/yahoo/vespa/athenz/identity/SiaBackedApacheHttpClient.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    @Override
+    public void close() {
```
Document somewhere (at least here) that the close is thread-safe.

bjorncs committed:
Misc improvements to close()

- Make close() idempotent
- Disallow execute() after close()
- Add redundant guards to refer()/release()

bjorncs commented:
@hakonhall PTAL
